File: crates/codex-plus-core/src/remote_mobile/live_source.js
(scope, binding, lease) => {
  const key = "__xuanMobileReplyObserver";
  const maxTextLength = 256 * 1024;
  const previous = window[key];
  if (previous?.binding === binding && typeof previous.renew === "function") {
    return previous.renew(scope, lease);
  }
  previous?.dispose();
  const state = {
    binding, scope: new Set(scope), items: new Map(), expires: Date.now() + lease,
    unsubscribers: [], dispatcher: null, timer: null,
  };
  const emit = (threadId, item) => {
    clearTimeout(item.timer);
    item.timer = null;
    if (!state.scope.has(threadId) || Date.now() > state.expires) return;
    try {
      window[binding](JSON.stringify({
        threadId, itemId: item.id, sequence: item.sequence, text: item.text,
        complete: item.complete,
      }));
      if (item.complete && state.items.get(threadId) === item) {
        state.items.delete(threadId);
      }
    } catch {
      state.dispose();
    }
  };
  const observe = (method, payload) => {
    const params = payload?.params || payload;
    const threadId = params?.threadId;
    if (!state.scope.has(threadId)) return;
    const id = params.itemId || params.item?.id;
    if (typeof id !== "string" || !id || id.length > 128) return;
    if (method !== "item/agentMessage/delta" && params.item?.type !== "agentMessage") return;
    let item = state.items.get(threadId);
    if (!item || item.id !== id) {
      // 中途加入时不能把缺少前缀的 delta 当作完整正文，等待完整消息事件或日志补齐。
      if (method === "item/agentMessage/delta") return;
      if (item?.timer) emit(threadId, item);
      // 新消息先锚定历史，后续只发送累计正文，桥接重放不会重复追加片段。
      item = { id, text: "", sequence: 0, complete: false, timer: null,
        suppressed: !!params.item?.channel && !["commentary", "final"].includes(params.item.channel) };
      state.items.set(threadId, item);
    }
    if (item.complete || item.suppressed) return;
    if (method === "item/agentMessage/delta") {
      if (typeof params.delta !== "string") return;
      if (item.text.length + params.delta.length > maxTextLength) {
        clearTimeout(item.timer);
        state.items.delete(threadId);
        return;
      }
      item.text += params.delta;
    } else if (typeof params.item.text === "string") {
      if (params.item.text.length > maxTextLength) {
        clearTimeout(item.timer);
        state.items.delete(threadId);
        return;
      }
      item.text = params.item.text;
    }
    item.complete = method === "item/completed";
    item.sequence++;
    if (method === "item/started" || item.complete) {
      emit(threadId, item);
    } else if (!item.timer) {
      item.timer = setTimeout(() => emit(threadId, item), 500);
    }
  };
  state.renew = (nextScope, nextLease) => {
    state.scope = new Set(nextScope);
    state.expires = Date.now() + nextLease;
    for (const id of state.items.keys()) {
      if (!state.scope.has(id)) {
        clearTimeout(state.items.get(id).timer);
        state.items.delete(id);
      }
    }
    return state.attach();
  };
  state.attach = () => {
    const dispatcher = window.__codexPlusRemoteSessionRecoveryDispatcher;
    if (!dispatcher || typeof dispatcher.subscribe !== "function") return false;
    if (state.dispatcher === dispatcher) return true;
    for (const unsubscribe of state.unsubscribers) {
      if (typeof unsubscribe === "function") unsubscribe();
    }
    state.unsubscribers = [];
    state.dispatcher = dispatcher;
    for (const method of ["item/started", "item/agentMessage/delta", "item/completed"]) {
      state.unsubscribers.push(dispatcher.subscribe(method, payload => observe(method, payload)));
    }
    return true;
  };
  state.dispose = () => {
    clearInterval(state.timer);
    for (const unsubscribe of state.unsubscribers) {
      if (typeof unsubscribe === "function") unsubscribe();
    }
    state.unsubscribers = [];
    for (const item of state.items.values()) clearTimeout(item.timer);
    state.items.clear();
    state.scope.clear();
    if (window[key] === state) delete window[key];
  };
  // CDP 异常断开时也会自动撤销观察者，不留下永久监听和正文缓存。
  state.timer = setInterval(() => {
    if (Date.now() > state.expires) state.dispose();
  }, Math.min(lease, 5000));
  window[key] = state;
  return state.attach();
}

File: crates/codex-plus-core/src/remote_mobile/live_source.rs
use std::collections::BTreeSet;
use std::time::Duration;

use anyhow::{Context, bail};
use futures_util::{SinkExt, StreamExt};
use serde::Deserialize;
use serde_json::{Value, json};
use tokio::sync::{mpsc, watch};
use tokio_tungstenite::tungstenite::Message;

const OBSERVER: &str = include_str!("live_source.js");
const BINDING: &str = "xuanMobileReplyDelta";

#[derive(Clone, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(super) struct ReplyEvent {
    pub thread_id: String,
    pub item_id: String,
    pub sequence: u64,
    pub text: String,
    pub complete: bool,
}

#[derive(Default)]
pub(super) struct ReplyOverlay {
    item_id: String,
    sequence: u64,
    base: String,
    text: String,
    complete: bool,
    retired: Vec<String>,
}

impl ReplyOverlay {
    pub fn apply(&mut self, event: ReplyEvent, history: &str) -> bool {
        if self.retired.contains(&event.item_id)
            || (self.item_id == event.item_id && event.sequence <= self.sequence)
        {
            return false;
        }
        if self.item_id != event.item_id {
            let base = self.render(history);
            if !self.item_id.is_empty() {
                self.retired.push(self.item_id.clone());
                if self.retired.len() > 32 {
                    self.retired.remove(0);
                }
            }
            self.base = base;
            self.item_id = event.item_id;
        }
        self.sequence = event.sequence;
        self.text = event.text;
        self.complete = event.complete;
        true
    }

    pub fn render(&self, history: &str) -> String {
        if self.text.is_empty() {
            return history.to_owned();
        }
        let projected = if self.base.is_empty() {
            self.text.clone()
        } else {
            format!("{}\n\n---\n\n{}", self.base, self.text)
        };
        // 日志追平或超前时以持久记录为准，否则保留正在生成且尚未落盘的尾部。
        if history.starts_with(&projected) {
            history.to_owned()
        } else if projected.starts_with(history) {
            projected
        } else if self.complete {
            history.to_owned()
        } else {
            projected
        }
    }

    pub fn pending(&self, history: &str) -> bool {
        self.render(history) != history
    }

    pub fn persisted(&self, item_id: &str) -> bool {
        !item_id.is_empty() && self.item_id == item_id
    }
}

pub(super) async fn run(
    mut scope: watch::Receiver<BTreeSet<String>>,
    output: mpsc::Sender<ReplyEvent>,
) {
    loop {
        let empty = scope.borrow().is_empty();
        if empty && scope.changed().await.is_err() {
            return;
        }
        if scope.borrow().is_empty() {
            continue;
        }
        // 只复用轩++已启动的本机调试端口，不扫描其他进程或改变桌面启动参数。
        #[cfg(not(test))]
        let port = crate::status::StatusStore::default()
            .load_latest()
            .ok()
            .flatten()
            .and_then(|status| status.debug_port);
        // 自动化测试只能连接显式构造的本机夹具，不能误连正在使用的桌面。
        #[cfg(test)]
        let port: Option<u16> = None;
        if let Some(port) = port {
            let _ = session(port, &mut scope, &output).await;
        }
        tokio::select! {
            _ = tokio::time::sleep(Duration::from_secs(5)) => {}
            changed = scope.changed() => if changed.is_err() { return; }
        }
    }
}

pub(super) fn set_scope(sender: &watch::Sender<BTreeSet<String>>, selected: &BTreeSet<String>) {
    sender.send_if_modified(|current| {
        if current == selected {
            false
        } else {
            current.clone_from(selected);
            true
        }
    });
}

async fn session(
    port: u16,
    scope: &mut watch::Receiver<BTreeSet<String>>,
    output: &mpsc::Sender<ReplyEvent>,
) -> anyhow::Result<()> {
    let targets = crate::cdp::list_targets(port).await?;
    let target = targets
        .iter()
        .find(|target| {
            crate::cdp::is_primary_codex_page_target(target) && target.url.starts_with("app://-/")
        })
        .context("桌面回复事件入口暂不可用")?;
    let url = target
        .web_socket_debugger_url
        .as_deref()
        .context("桌面回复连接暂不可用")?;
    crate::cdp::validate_cdp_websocket_url(url, port)?;
    let (mut socket, _) = tokio::time::timeout(
        Duration::from_secs(5),
        tokio_tungstenite::connect_async(url),
    )
    .await??;
    let mut command_id = 1_u64;
    super::send(
        &mut socket,
        &json!({"id": command_id, "method": "Runtime.addBinding",
        "params": {"name": BINDING}}),
    )
    .await?;
    let mut lease = tokio::time::interval(Duration::from_secs(5));
    lease.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
    let result = async {
        loop {
            tokio::select! {
                changed = scope.changed() => {
                    if changed.is_err() || scope.borrow().is_empty() { break; }
                    lease.reset_immediately();
                }
                _ = lease.tick() => {
                    let selected = serde_json::to_string(&*scope.borrow())?;
                    command_id += 1;
                    let expression = format!("({OBSERVER})({selected}, {binding}, 15000)",
                        binding = serde_json::to_string(BINDING)?);
                    super::send(&mut socket, &json!({"id": command_id, "method": "Runtime.evaluate",
                        "params": {"expression": expression, "returnByValue": true}})).await?;
                }
                message = socket.next() => {
                    match message {
                        Some(Ok(Message::Text(text))) => {
                            let value: Value = serde_json::from_str(&text)?;
                            if value["error"].is_object() { bail!("桌面回复观察暂不可用"); }
                            if value["method"] != "Runtime.bindingCalled" || value["params"]["name"] != BINDING {
                                continue;
                            }
                            let event: ReplyEvent = serde_json::from_str(
                                value["params"]["payload"].as_str().context("回复事件格式无效")?
                            )?;
                            if !scope.borrow().contains(&event.thread_id) || event.item_id.is_empty()
                                || event.item_id.len() > 128 || event.sequence == 0
                            {
                                continue;
                            }
                            // 累计正文可覆盖队列中旧片段；队列拥堵时仍由最终日志补齐。
                            let _ = output.try_send(event);
                        }
                        Some(Ok(Message::Ping(bytes))) => socket.send(Message::Pong(bytes)).await?,
                        Some(Ok(Message::Close(_))) | None => break,
                        Some(Err(error)) => return Err(error.into()),
                        _ => {}
                    }
                }
            }
        }
        anyhow::Ok(())
    }.await;
    let cleanup = async {
        super::send(
            &mut socket,
            &json!({"id": command_id + 1, "method": "Runtime.evaluate",
            "params": {"expression": "window.__xuanMobileReplyObserver?.dispose()"}}),
        )
        .await?;
        super::send(
            &mut socket,
            &json!({"id": command_id + 2, "method": "Runtime.removeBinding",
            "params": {"name": BINDING}}),
        )
        .await?;
        socket.close(None).await?;
        anyhow::Ok(())
    };
    let _ = tokio::time::timeout(Duration::from_secs(2), cleanup).await;
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    fn event(item: &str, sequence: u64, text: &str, complete: bool) -> ReplyEvent {
        ReplyEvent {
            thread_id: "fixture_task_0001".into(),
            item_id: item.into(),
            sequence,
            text: text.into(),
            complete,
        }
    }

    #[test]
    fn live_delta_is_visible_before_log_and_final_history_does_not_duplicate() {
        let mut overlay = ReplyOverlay::default();
        assert!(overlay.apply(event("item1", 1, "", false), "历史"));
        overlay.apply(event("item1", 2, "正在", false), "历史");
        assert_eq!(overlay.render("历史"), "历史\n\n---\n\n正在");
        assert!(!overlay.apply(event("item1", 1, "旧", false), "历史"));
        overlay.apply(event("item1", 3, "正在生成", true), "历史");
        let full = "历史\n\n---\n\n正在生成";
        assert_eq!(overlay.render("历史"), full);
        assert_eq!(overlay.render(full), full);
        assert!(!overlay.pending(full));
        overlay.apply(event("item2", 1, "正在生成", true), full);
        assert_eq!(overlay.render(full), format!("{full}\n\n---\n\n正在生成"));
        assert!(!overlay.apply(event("item1", 4, "旧消息", true), full));
        assert!(overlay.persisted("item2"));
    }

    #[tokio::test]
    async fn cdp_source_delivers_events_and_unsubscribes_without_executing_commands() {
        use tokio::io::{AsyncReadExt, AsyncWriteExt};
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let port = listener.local_addr().unwrap().port();
        let server = tokio::spawn(async move {
            let (mut http, _) = listener.accept().await.unwrap();
            let mut request = vec![0_u8; 4096];
            let read = http.read(&mut request).await.unwrap();
            assert!(String::from_utf8_lossy(&request[..read]).starts_with("GET /json "));
            let body = json!([{"id": "fixture", "type": "page", "title": "Codex",
                "url": "app://-/index.html",
                "webSocketDebuggerUrl": format!("ws://127.0.0.1:{port}/devtools/page/fixture")}])
            .to_string();
            http.write_all(format!("HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}", body.len()).as_bytes()).await.unwrap();
            http.shutdown().await.unwrap();
            let (tcp, _) = listener.accept().await.unwrap();
            let mut socket = tokio_tungstenite::accept_async(tcp).await.unwrap();
            let mut delivered = false;
            let mut removed = false;
            while let Some(message) = socket.next().await {
                match message.unwrap() {
                    Message::Text(text) => {
                        let command: Value = serde_json::from_str(&text).unwrap();
                        assert!(matches!(
                            command["method"].as_str().unwrap(),
                            "Runtime.addBinding" | "Runtime.evaluate" | "Runtime.removeBinding"
                        ));
                        if command["method"] == "Runtime.evaluate" && !delivered {
                            assert!(
                                command["params"]["expression"]
                                    .as_str()
                                    .unwrap()
                                    .contains("fixture_task_0001")
                            );
                            let event = json!({"threadId":"fixture_task_0001","itemId":"item1",
                                "sequence":1,"text":"真实传输夹具","complete":false});
                            let notification = json!({"method":"Runtime.bindingCalled",
                                "params":{"name":BINDING,"payload":event.to_string()}});
                            socket
                                .send(Message::Text(notification.to_string().into()))
                                .await
                                .unwrap();
                            delivered = true;
                        }
                        if command["method"] == "Runtime.removeBinding" {
                            removed = true;
                        }
                    }
                    Message::Close(_) => break,
                    _ => {}
                }
            }
            assert!(delivered && removed);
        });
        let (scope_tx, mut scope) =
            watch::channel(BTreeSet::from(["fixture_task_0001".to_owned()]));
        let (output, mut events) = mpsc::channel(4);
        let client = tokio::spawn(async move { session(port, &mut scope, &output).await });
        let received = tokio::time::timeout(Duration::from_secs(10), events.recv())
            .await
            .unwrap()
            .unwrap();
        assert_eq!(received.text, "真实传输夹具");
        scope_tx.send_replace(BTreeSet::new());
        tokio::time::timeout(Duration::from_secs(10), client)
            .await
            .unwrap()
            .unwrap()
            .unwrap();
        tokio::time::timeout(Duration::from_secs(10), server)
            .await
            .unwrap()
            .unwrap();
    }
}
